src/Azure/Orleans.Transactions.AzureStorage/TransactionalState/AzureTableTransactionalStateStorage.cs

```diff
@@ -0,0 +1,184 @@
+        {
+            this.table = table;
+            this.partition = partition;
+            this.JsonSettings = JsonSettings;
+            this.logger = logger;
+        }
+
+        public async Task<TransactionalStorageLoadResponse<TState>> Load()
+        {
+            try
+            {
+                Task<KeyEntity> keyTask = ReadKey();
+                Task<List<StateEntity>> statesTask = ReadStates();
+                this.key = await keyTask;
+                this.states = await statesTask;
+                if (string.IsNullOrEmpty(this.key.ETag))
+                {
+                    return new TransactionalStorageLoadResponse<TState>();
+                }
+                TState commitedState = TryFindState(this.key.CommittedTransactionId, out TState found) ? found : new TState();
+                var pendingStates = states.Select(s => new PendingTransactionState<TState>(s.TransactionId, s.SequenceId, s.GetState<TState>(this.JsonSettings))).ToList();
+                return new TransactionalStorageLoadResponse<TState>(this.key.ETag, commitedState, this.key.Metadata, pendingStates);
+            } catch(Exception ex)
+            {
+                this.logger.LogError("Transactional state load failed {Exception}.", ex);
+                throw;
+            }
+        }
+
+        public async Task<string> Persist(string expectedETag, string metadata, List<PendingTransactionState<TState>> statesToPrepare)
+        {
+            try
+            {
+                var batchOperation = new TableBatchOperation();
+
+                this.key.ETag = expectedETag;
+                this.key.Metadata = metadata;
+                if (string.IsNullOrEmpty(this.key.ETag))
+                    batchOperation.Insert(this.key);
+                else
+                    batchOperation.Replace(this.key);
+
+                // add new states
+                List<Tuple<string,long>> stored = this.states.Select(s => Tuple.Create(s.TransactionId, s.SequenceId)).ToList();
+                List<StateEntity> newStates = new List<StateEntity>();
+                foreach (PendingTransactionState<TState> pendingState in statesToPrepare.Where(p => !stored.Contains(Tuple.Create(p.TransactionId, p.SequenceId))))
+                {
+                    var newState = StateEntity.Create(this.JsonSettings, this.partition, pendingState);
+                    newStates.Add(newState);
+                    batchOperation.Insert(newState);
+                }
+
+                if (batchOperation.Count > AzureTableConstants.MaxBatchSize)
+                {
+                    this.logger.LogError("Too many pending states. PendingStateCount {PendingStateCount}.", batchOperation.Count);
+                    throw new InvalidOperationException($"Too many pending states. PendingStateCount {batchOperation.Count}");
```

Comment:
To make this robust, should probably prepare using multiple batches instead of throwing an exception. Not sure if that level of reliability is necessary, I believe this provider is not meant for production use?

Reply:
This provider is intended for production use, but system is still in development.  I'll add issues to revisit shortcomings.

Operation must be atomic, which is only provided by a single batch.  We could change api surface to support partial success, but given current api expectations and storage capabilities, best we can do is enforce this limitation.

Reply:
Ok, probably good enough for now. We may revisit this when looking at the storage interface for the distributed TM.